src/imgui/Shortcuts.hh

```diff
@@ -0,0 +1,82 @@
+#ifndef IMGUI_SHORTCUTS_HH
+#define IMGUI_SHORTCUTS_HH
+
+#include "ImGuiUtils.hh"
+
+#include "strCat.hh"
+#include "imgui_stdlib.h"
+
+#include <functional>
+#include <optional>
+#include <string_view>
+#include <vector>
+
+namespace openmsx {
+
+class ImGuiManager;
+class GlobalSettings;
+class SettingsConfig;
+
+enum ShortcutIndex {
+	GOTO_MEMORY_ADDRESS = 0,
+	GOTO_DISASM_ADDRESS,
+	NUM_SHORTCUTS,
+};
+
+enum ShortcutType {
+	LOCAL,    // if ImGui widget has focus
+	GLOBAL,   // inside ImGui layer only
+};
+
+class Shortcuts final
+{
+public:
+	struct Data {
+		ShortcutIndex index;
+		ImGuiKeyChord keychord;
+		bool local = false;
+		bool repeat = false;
+	};
+
+	// userlevel shortcut type
+	using KeyChordSet = std::vector<ImGuiKeyChord>;   // unsorted
+
+	Shortcuts(const Shortcuts&) = delete;
+	Shortcuts& operator=(const Shortcuts&) = delete;
+	explicit Shortcuts(SettingsConfig& config_);
+
+	// Shortcuts
+	static std::string getShortcutName(ShortcutIndex index);
+	Shortcuts::Data& getShortcut(ShortcutIndex index);
+	void setShortcut(ShortcutIndex index, ImGuiKeyChord keychord, std::optional<bool> local = {}, std::optional<bool> repeat = {});
+	bool isPressed(ShortcutIndex index);
+
+	template<typename XmlStream>
+	void saveShortcuts(XmlStream& xml) const
+	{
+		xml.begin("shortcuts");
+		for (const auto& data : shortcuts) {
+			xml.begin("shortcut");
+			if (data.keychord != ImGuiKey_None) {
+				xml.attribute("keychord", std::to_string(static_cast<int>(data.keychord)));
+			}
+			if (data.repeat) xml.attribute("repeat", "true");
+			if (data.local) xml.attribute("local", "true");
+			xml.data(std::to_string(static_cast<int>(data.index)));
```

Comment:
Both uses of `int -> string` (lines 61 and 65) are wrong. Or at least not robust.

Suppose you saved `settings.xml`, then upgrade your openMSX version, then load `settings.xml` again. You'd expect this to setup the same shortcuts again. Though when values were added/removed in `enum ShortcutIndex` this will give unexpected results. And the same (but maybe less frequent) for `ImGuiKeyChord`. We cannot assume that the numerical values of these enums remain stable over releases.

Instead we should save (and parse) string-representations of these enums. For `ImGuiKeyChord` we already have a `getKeyChordName()`, but we'd also need a corresponding parsing function.

For `ShortcutIndex` we might reuse the strings from `getShortcutName()`. My original idea was that we could also use this string as a full description of the shortcut (the strings that are now hardcoded in `ImGuiSettings.cc` lines 437 and 442). But those probably aren't suited for this (you don't want to invalidate settings.xml after fixing typos in the (possibly long) description string).

Reply:
> Instead we should save (and parse) string-representations of these enums. For `ImGuiKeyChord` we already have a `getKeyChordName()`, but we'd also need a corresponding parsing function.

We should not be using GetKeyChordName because the ImGui::GetKeyName that it uses comes with this caveat:
```    
IMGUI_API const char*   GetKeyName(ImGuiKey key);                                           // [DEBUG] returns English name of the key. Those names a provided for debugging purpose and are not meant to be saved persistently not compared.
```
So I changed it to use SDL_KeyEvent keycodes to strings conversion that SDL already offers.